### PR TITLE
Update docBlock of get_option() to better explain the expected return…

### PR DESCRIPTION
**Pull request by rene-hermenau**
Update docBlock of get_option() to better explain the expected return values, taking into account all the exceptions like using $default or the internal filters

Trac ticket: https://core.trac.wordpress.org/ticket/51278

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/option.php [hunk 1]
```diff
@@ -16,8 +16,40 @@
  *
  * If the option was serialized then it will be unserialized when it is returned.
  *
- * Any scalar values will be returned as strings. You may coerce the return type of
- * a given option by registering an {@see 'option_$option'} filter callback.
+ * Any scalar values will be returned as strings!
```
Comment (azaozz) [inline src/wp-includes/option.php:19]: A string is a "scalar value" too.

Comment (azaozz) [reply src/wp-includes/option.php:19]:
```suggestion
 * Non-string scalar values will be returned as strings when the options are retrieved from the database and the values are not filtered.
```

File: src/wp-includes/option.php [hunk 2]
```diff
@@ -16,8 +16,40 @@
...
+ *
+ * Exception: It does return scalar values for a value that doesn't originate from the database storage.
+ * for example $default is scalar and the option name does not exist or the return value from any of its filters is scalar
+ * (pre_option_{$option}, default_option_{$option}, and option_{$option}),
+ * the return value of $default will still be scalar and not converted to a string!
```
Comment (azaozz) [inline src/wp-includes/option.php:24]:
```suggestion
 *
 * Exceptions: 
 * 1. When the option is not saved in the database the default value is returned, if provided. If not, boolean `false` is returned.
 * 2. When one of the Options API filters is used:
 * {@see pre_option_{$option}}, {@see default_option_{$option}}, and {@see option_{$option}},
 * the return value may not match the expected type.
```

File: src/wp-includes/option.php [hunk 3]
```diff
@@ -16,8 +16,40 @@
...
+ * Storing database options like this:
+ *
+ * add_option( 'option_name__false', false );
+ * add_option( 'option_name__true', true );
+ * add_option( 'option_name__0', 0 );
+ * add_option( 'option_name__1', 1 );
+ * add_option( 'option_name__str_0', '0' );
+ * add_option( 'option_name__str_1', '1' );
+ *
+ * will return with using get_option('option_name__[]'):
```
Comment (azaozz) [inline src/wp-includes/option.php:37]:
```suggestion
 * get_option( 'option_name__[]' ) will return the following:
```

File: src/wp-includes/option.php [hunk 4]
```diff
@@ -16,8 +16,40 @@
...
+ * Any scalar values will be returned as strings!
+ *
+ * Exception: It does return scalar values for a value that doesn't originate from the database storage.
+ * for example $default is scalar and the option name does not exist or the return value from any of its filters is scalar
+ * (pre_option_{$option}, default_option_{$option}, and option_{$option}),
+ * the return value of $default will still be scalar and not converted to a string!
+ *
+ * Examples:
+ *
+ * Storing database options like this:
```
Comment (azaozz) [inline src/wp-includes/option.php:28]:
```suggestion
 * When adding options like this:
```

File: src/wp-includes/option.php [hunk 5]
```diff
@@ -16,8 +16,40 @@
...
+ * string(1) "0"
+ * string(1) "1"
+ *
+ * Example of returning scalar value if option does not exist:
+ *
+ * get_option('option_name_not_exist', 1);
+ *
+ * Will return:
+ *
+ * (int) 1
```
Comment (azaozz) [inline src/wp-includes/option.php:52]: Thinking this is not needed. Seems clear enough above :)

**Comment by azaozz** [comment]
Added some suggestions/changes. Also may be good to mention that as non-scalar option values are serialized before saving, they remain unchanged. For example adding `array( true, false, 0, 1 )` will return the exact same array.

**Comment by rene-hermenau** [comment]
> Added some suggestions/changes. Also may be good to mention that as non-scalar option values are serialized before saving, they remain unchanged. For example adding `array( true, false, 0, 1 )` will return the same array.

@azaozz  Excellent points! I feel ashamed for overlooking that strings are scalar. You've passed my test:-)

 `get_option()` is a very tricky method with all these possible different types of returns.  We are right on time to work out this improvement. I've added your suggestions. Please have a look at them.

**Pull request closed without merging**